File: 자료구조/BOJ_1918.js
const input = require('fs').readFileSync('/dev/stdin').toString().trim();
var postfix = '';
var stack = [];

const op = {'+' : 1, '-' : 1, '*' : 2, '/' : 2, '(' : 0};

for(w of input){
    if(w >= 'A' && w <= 'Z'){
        postfix += w;
    } else if (w === '('){
        stack.push(w)
    } else if (w == ')'){
        while(stack.length > 0 && stack.at(-1) !== '('){
            postfix += stack.pop();
        }
        stack.pop()
    } else {
        while(stack.length > 0 && op[w] <= op[stack.at(-1)]){
            postfix += stack.pop();
        }
        stack.push(w)
    }
}

while(stack.length > 0){
    postfix += stack.pop();
}

console.log(postfix)

File: 자료구조/BOJ_1918.py
# 후위 표기식, G2
from sys import stdin

arr = stdin.readline().rstrip()

op = {'+' : 1, '-' : 1, '*' : 2, '/' : 2, '(' : 0}
postfix = []
stack = []

res = ''

for w in arr:
    if 'A' <= w <='Z':
        postfix.append(w)
    elif w == '(':
        stack.append(w)
    elif w == ')':
        while stack and stack[-1] != '(':
            postfix.append(stack.pop())
        stack.pop()
    else:
        while stack and op[w] <= op[stack[-1]]:
            postfix.append(stack.pop())
        stack.append(w)

while len(stack) != 0:
    postfix.append(stack.pop())

for val in postfix:
    res += val
print(res)